lib/host/wasi/inode-win.cpp

```diff
@@ -82,53 +434,310 @@ INode INode::stdErr() noexcept {
+    uint64_t Aticks = ATim / NANOSECONDS_PER_TICK + TICKS_TO_UNIX_EPOCH;
+    AFileTime.dwLowDateTime = static_cast<DWORD>(Aticks % 0x100000000ULL);
+    AFileTime.dwHighDateTime = static_cast<DWORD>(Aticks / 0x100000000ULL);
+  } else if (FstFlags & __WASI_FSTFLAGS_ATIM_NOW) {
+    GetSystemTimeAsFileTime(&AFileTime);
+  }
+
+  // For setting modification time
+  if (FstFlags & __WASI_FSTFLAGS_MTIM) {
+    uint64_t Mticks = MTim / NANOSECONDS_PER_TICK + TICKS_TO_UNIX_EPOCH;
+    MFileTime.dwLowDateTime = static_cast<DWORD>(Mticks % 0x100000000ULL);
+    MFileTime.dwHighDateTime = static_cast<DWORD>(Mticks / 0x100000000ULL);
+  } else if (FstFlags & __WASI_FSTFLAGS_MTIM_NOW) {
+    GetSystemTimeAsFileTime(&MFileTime);
+  }
+
+  if (unlikely(SetFileTime(Handle, nullptr, &AFileTime, &MFileTime)) == FALSE) {
+    return WasiUnexpect(fromWinError(GetLastError()));
+  }
+
+  return {};
 }
 
-WasiExpect<void> INode::fdPread(Span<Span<uint8_t>>, __wasi_filesize_t,
-                                __wasi_size_t &) const noexcept {
-  return WasiUnexpect(__WASI_ERRNO_NOSYS);
+WasiExpect<void> INode::fdPread(Span<Span<uint8_t>> IOVs,
+                                __wasi_filesize_t Offset,
+                                __wasi_size_t &NRead) const noexcept {
+  NRead = 0;
+  uint64_t LocalOffset = Offset;
+
+  for (auto IOV : IOVs) {
+    DWORD NumberOfBytesRead = 0;
+    OVERLAPPED Result;
+
+    Result.Offset = static_cast<uint32_t>(LocalOffset);
+    Result.OffsetHigh = static_cast<uint32_t>(LocalOffset >> 32);
+
+    // There maybe issues due to casting IOV.size() to unit32_t
+    BOOL ReadResult =
+        ReadFile(Handle, IOV.data(), static_cast<uint32_t>(IOV.size()),
+                 &NumberOfBytesRead, &Result);
+
+    if (GetLastError() == ERROR_IO_PENDING) {
+      // Wait for the Write to complete
+      if (unlikely(GetOverlappedResult(Handle, &Result, &NumberOfBytesRead,
+                                       TRUE)) == FALSE) {
+        return WasiUnexpect(fromWinError(GetLastError()));
+      }
+    } else if (unlikely(ReadResult == FALSE)) {
+      return WasiUnexpect(fromWinError(GetLastError()));
+    }
+    LocalOffset += NumberOfBytesRead;
+    NRead += NumberOfBytesRead;
+  }
+
+  return {};
 }
 
-WasiExpect<void> INode::fdPwrite(Span<Span<const uint8_t>>, __wasi_filesize_t,
-                                 __wasi_size_t &) const noexcept {
-  return WasiUnexpect(__WASI_ERRNO_NOSYS);
+WasiExpect<void> INode::fdPwrite(Span<Span<const uint8_t>> IOVs,
+                                 __wasi_filesize_t Offset,
+                                 __wasi_size_t &NWritten) const noexcept {
+  NWritten = 0;
+  uint64_t LocalOffset = Offset;
+
+  for (auto IOV : IOVs) {
+    DWORD NumberOfBytesWritten = 0;
+    OVERLAPPED Result;
+
+    Result.Offset = static_cast<uint32_t>(LocalOffset);
+    Result.OffsetHigh = static_cast<uint32_t>(LocalOffset >> 32);
+
+    // There maybe issues due to casting IOV.size() to unit32_t
```

Comment:
What's the issue?

Reply:
Same as above.
> `IOV.size()` returns a 64 bit integer but the underling windows API uses a 32 bit integer. So can there be any issues due to the narrowing of the range?

Reply:
Well, just add a check here, if the range is larger than the limit of windows API, report an error.

Reply:
Okay

Reply:
Instead should I divide the larger number into two parts - `INT_MAX` and `IOV.size() - INT_MAX` and then perform two operations with these?